File: Object-C/BlackFireIM/Example/Lib/MSIMSDK-UIKit/VoiceKit/MSVoipCenter.m
//
//  MSVoipCenter.m
//  BlackFireIM
//
//  Created by benny wang on 2022/3/2.
//

#import "MSVoipCenter.h"
#import <MSIMSDK/MSIMSDK.h>
#import "MSCallManager.h"
#import <CallKit/CallKit.h>
#import "MSCallViewController.h"

@interface MSVoipCenter()

@property(nonatomic,strong) NSMutableDictionary *uuids;

@property(nonatomic,strong) CXCallController *callVC;

/// 记录正在通话的calling
@property(nonatomic,copy) NSString *currentCalling;

@end
@implementation MSVoipCenter

+ (instancetype)shareInstance
{
    static dispatch_once_t onceToken;
    static MSVoipCenter * g_sharedInstance = nil;
    dispatch_once(&onceToken, ^{
        g_sharedInstance = [[MSVoipCenter alloc] init];
    });
    return g_sharedInstance;
}

- (instancetype)init
{
    if (self = [super init]) {
        _uuids = [NSMutableDictionary dictionary];
        [[NSNotificationCenter defaultCenter]addObserver:self selector:@selector(recieveNeedToDismissVoipView:) name:@"kRecieveNeedToDismissVoipView" object:nil];
    }
    return self;
}

- (NSString *)createUUIDWithRoomID:(NSString *)room_id fromUid:(NSString *)fromUid subType:(MSCallType)callType
{
    NSString *uuid = [NSUUID UUID].UUIDString;
    self.uuids[uuid] = @{@"room_id": room_id,@"from": fromUid,@"call_type": @(callType)};
    self.callVC = [[CXCallController alloc]initWithQueue:dispatch_get_main_queue()];
    return uuid;
}

- (NSString *)roomIDWithUUID:(NSString *)uuid
{
    NSDictionary *dic = self.uuids[uuid];
    return dic[@"room_id"];
}

- (void)acceptCallWithUuid:(NSString *)uuid
{
    if (uuid.length == 0) return;
    NSDictionary *dic = self.uuids[uuid];
    if (dic == nil) return;
    MSCallType type = [dic[@"call_type"]integerValue];
    NSString *room_id = dic[@"room_id"];
    self.currentCalling = uuid;
    [[MSCallManager shareInstance] recieveCall:dic[@"from"] creator:dic[@"from"] callType:type action:CallAction_Call room_id:room_id];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [[MSCallManager shareInstance]autoAcceptVoipCall:type room_id:room_id];
    });
}

- (void)endCallWithUuid:(NSString *)uuid
{
    if (uuid.length == 0) return;
    NSDictionary *dic = self.uuids[uuid];
    if (dic == nil) return;
    MSCallType type = [dic[@"call_type"]integerValue];
    NSString *room_id = dic[@"room_id"];
    if (self.currentCalling == nil) {
        [[MSCallManager shareInstance] sendMessageType:CallAction_Reject option:IMCUSTOM_SIGNAL room_id:room_id toReciever:dic[@"from"]];
    }else {
        [[MSCallManager shareInstance].callVC recieveVoipEnd:type room_id:room_id];
    }
    self.currentCalling = nil;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [[NSNotificationCenter defaultCenter] postNotificationName:UIApplicationDidEnterBackgroundNotification object:nil];
    });
}

- (void)muteCall:(BOOL)isMute uuid:(NSString *)uuid
{
    if (uuid.length == 0) return;
    NSDictionary *dic = self.uuids[uuid];
    if (dic == nil) return;
//    MSCallType type = [dic[@"call_type"]integerValue];
}

- (void)recieveNeedToDismissVoipView:(NSNotification *)note
{
    self.currentCalling = nil;
    for (NSString *uuid in self.uuids.allKeys) {
        [self.uuids removeObjectForKey:uuid];
        CXEndCallAction *action = [[CXEndCallAction alloc]initWithCallUUID:[[NSUUID alloc]initWithUUIDString:uuid]];
        CXTransaction *transaction = [[CXTransaction alloc]initWithAction:action];
        [self.callVC requestTransaction:transaction completion:^(NSError * _Nullable error) {
            MSLog(@"error: %@",error);
        }];
    }
}

- (void)didActivateAudioSession
{
    [[MSCallManager shareInstance].callVC didActivateAudioSession];
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/CallKit/MSVoipCenter.swift
//
//  MSVoipCenter.swift
//  MSIM-Swift
//
//  Created by benny wang on 2022/3/4.
//

import UIKit
import CallKit
import AgoraRtcKit
import MSIMSDK


struct CallItem {
    var room_id: String
    var from: String
    var call_type: MSCallType
}
class MSVoipCenter: NSObject,AgoraRtcEngineDelegate {
    
    static let shared: MSVoipCenter = MSVoipCenter()
    
    var callVC: CXCallController?
    
    var uuids: [String: CallItem] = [:]
    
    private(set) var currentCalling: String?
    
    private override init() {
        super.init()
        NotificationCenter.default.addObserver(self, selector: #selector(recieveNeedToDismissVoipView), name: NSNotification.Name.init("kRecieveNeedToDismissVoipView"), object: nil)
    }
    
    func createUUIDWithRoomID(room_id: String,fromUid: String,callType: MSCallType) -> String {
        
        let uuid = UUID().uuidString
        self.uuids[uuid] = CallItem(room_id: room_id, from: fromUid, call_type: callType)
        self.callVC = CXCallController(queue: .main)
        return uuid
    }
    
    func acceptCallWithUuid(uuid: String) {
        if let item = self.uuids[uuid] {
            self.currentCalling = uuid
            MSCallManager.shared.recieveCall(from: item.from, creator: item.from, callType: item.call_type, action: .call, room_id: item.room_id)
            
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                MSCallManager.shared.autoAcceptVoipCall(type: item.call_type,room_id: item.room_id)
            }
        }
    }
    
    func endCallWithUuid(uuid: String) {
        if let item = self.uuids[uuid] {
            if self.currentCalling == nil {
                MSCallManager.shared.callToPartner(partner_id: item.from, creator: item.from, callType: item.call_type, action: .reject, room_id: item.room_id)
            }else {
                MSCallManager.shared.callVC.recieveVoipEnd(type: item.call_type,room_id: item.room_id)
            }
            self.currentCalling = nil
            DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
                NotificationCenter.default.post(name: UIApplication.didEnterBackgroundNotification, object: nil)
            }
        }
    }
    
    func muteCall(isMute: Bool, uuid: String) {
        
    }
    
    @objc func recieveNeedToDismissVoipView(note: Notification) {
        
        self.currentCalling = nil
        for uuid in self.uuids.keys {
            self.uuids.removeValue(forKey: uuid)
            let action = CXEndCallAction.init(call: UUID(uuidString: uuid)!)
            let transaction = CXTransaction(action: action)
            self.callVC?.request(transaction, completion: { _ in
                
            })
        }
    }
    
    func didActivateAudioSession() {
        MSCallManager.shared.callVC.didActivateAudioSession()
    }
}
